src/datasets/arrow_dataset.py

```diff
@@ -5111,10 +5114,11 @@ def shards_with_embedded_external_files(shards):
             shards = shards_with_embedded_external_files(shards)
 
         files = hf_api_list_repo_files(api, repo_id, repo_type="dataset", revision=branch, use_auth_token=token)
-        data_files = [file for file in files if file.startswith("data/")]
+        data_dir = f"{config_name}/data" if config_name != "default" else "data"  # for backward compatibility
```

Comment:
Maybe the pattern can be switched to
```python
SPLIT_PATTERN_SHARDED = "{split}-[0-9][0-9][0-9][0-9][0-9]-of-[0-9][0-9][0-9][0-9][0-9]*.*"
```

Since by default the `data_dir` would be `"data"`, the pattern would be executed inside this directory and it should work

Reply:
then it would recognize (example for "train") `"data/train"` as a split during pattern search because it searches for `"*-[0-9]....."` [here](https://github.com/huggingface/datasets/blob/main/src/datasets/data_files.py#L224) and then the pattern becomes "data/data/train-[0-9]....." [here](https://github.com/huggingface/datasets/blob/main/src/datasets/data_files.py#L483) and it ended up with an error in the end (`Unable to resolve any data file that matches 'data/train-[0-9]....' ...`). 

I spend some time trying to fix the regex but didn't succeed...

Reply:
Oh I see. It's because `_resolve_single_pattern_locally` and `_resolve_single_pattern_in_dataset_repository` don't return paths relative to the input `base_path` I think. Could you try modifying them ?

Otherwise I'm happy to do it, since I added them in the first place

Reply:
@lhoestq didn't manage to make it work as you suggested so I left `data_dir` in metadata as is for now